Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c

```diff
@@ -81,80 +80,116 @@ pas_allocation_result pas_probabilistic_guard_malloc_allocate(pas_large_heap* la
     pas_allocation_result result = pas_allocation_result_create_failure();
 
     if (verbose)
-        pas_log("Memory requested to allocate %zu\n", size);
+        pas_log("Memory requested to allocate %zu with alignment %zu\n", size, alignment);
 
     if (!large_heap || !size || !heap_config || !transaction)
         return result;
 
     const size_t page_size = pas_page_malloc_alignment();
 
     type = pas_heap_for_large_heap(large_heap)->type;
-    alignment = heap_config->get_type_alignment(type);
+    alignment = PAS_MAX(alignment, heap_config->get_type_alignment(type));
     alignment = PAS_MAX(alignment, heap_config->large_alignment);
+    PAS_ASSERT(page_size >= heap_config->large_alignment);
+
+    size_t mem_to_waste = 0;
+    size_t mem_to_alloc = 0;
+    if (alignment >= page_size) {
+        /*
+         * In this case, slide becomes always n * page_size since alignment is n * page_size.
+         * Also, to achieve this alignment request, we need this expanded size, and this is not something wasted for PGM.
+         * So we set 0 to mem_to_waste.
+         * The worst scenario is the allocated pointer is off by one page from the alignment boundary. In that case, we need
+         * to add (alignment - page_size) to obtain the next alignment boundary. This means we need to allocate at least
+         * size + (alignment - page_size) memory to find alignment boundary with size.
+         */
+        size_t size_with_alignment_reservation = pas_round_up(size + alignment - page_size, page_size);
+        mem_to_waste = 0;
+        mem_to_alloc = (2 * page_size) + size_with_alignment_reservation;
+    } else {
+        /*
+         * Since both page_size and alignment are power-of-two and alignment is smaller than page_size,
+         * page_size is always n * alignment. Since size_with_alignment_reservation is m * alignment,
+         * mem_to_waste is also x * alignment. Thus right-align will be guaranteed to be aligned.
+         */
+        size_t size_with_alignment_reservation = pas_round_up(size, alignment);
+        mem_to_waste = (page_size - (size_with_alignment_reservation % page_size)) % page_size;
```

Comment:
Recording my reasoning here for posterity: The only time `% page_size` is interesting is if `(size_with_alignment_reservation % page_size)` == `page_size`.   Though we know that `alignment < page_size`, we don't know what `size` is. Hence, `size_with_alignment_reservation` can be exactly an even number of page, which leads:
    `(size_with_alignment_reservation % page_size)` == `0`
    ==> `(page_size - (size_with_alignment_reservation % page_size))` == `page_size`

... the waste here is 0, not `page_size`.  Hence, the `% page_size` is needed.